qiskit/algorithms/gibbs_state_preparation/gibbs_state_sampler.py

```diff
@@ -0,0 +1,251 @@
+        Raises:
+            ValueError: If ansatz and ansatz_params_dict are not both provided.
+        """
+        if not self._ansatz or not self._ansatz_params_dict:
+            raise ValueError(
+                "Both ansatz and ansatz_params_dict must be present in the class to compute "
+                "gradients."
+            )
+        gradient_params = list(self.ansatz_params_dict.keys())
+
+        operator = CircuitStateFn(self.ansatz)
+        # combo fn for an operator here, telling how to batch results together, pass combo fn to
+        # Gradient
+        # state_grad_with_aux_regs = Gradient(grad_method=gradient_method).convert(
+        #     operator=operator, params=gradient_params
+        # )
+        # sampler = CircuitSampler(backend=backend).convert(
+        #     state_grad_with_aux_regs, self._ansatz_params_dict
+        # )
+        # gradient_amplitudes_with_aux_regs = sampler.eval()[0]
+        gradient_amplitudes_with_aux_regs = Gradient(grad_method=gradient_method).gradient_wrapper(
+            operator, self.ansatz.ordered_parameters, backend=backend
+        )
+        # Get the values for the gradient of the sampling probabilities w.r.t. the Ansatz parameters
+        gradient_amplitudes_with_aux_regs = gradient_amplitudes_with_aux_regs(
+            self.ansatz_params_dict.values()
+        )
+        # TODO gradients of amplitudes or probabilities?
+        state_grad = self._discard_aux_registers_gradients(
+            gradient_amplitudes_with_aux_regs, self.ansatz.num_qubits
+        )
+        return state_grad
+
+    def calc_hamiltonian_gradients(
+        self,
+        backend: Union[Backend, BaseBackend, QuantumInstance],
+        gradient_method: str = "param_shift",
+    ) -> Dict[Parameter, NDArray[Union[complex, float]]]:
+        """
+        Calculates gradients of the visible part of a Gibbs state w.r.t. parameters of a
+        Hamiltonian that gave rise to the Gibbs state.
+        Args:
+            backend: A backend used for sampling circuit probabilities.
+            gradient_method: A desired gradient method chosen from the Qiskit Gradient Framework.
+        Returns:
+            Calculated gradients of the visible part of a Gibbs state w.r.t. parameters of a
+            Hamiltonian that gave rise to the Gibbs state.
+        """
+        ansatz_gradients = self.calc_ansatz_gradients(backend, gradient_method)
+        gibbs_state_hamiltonian_gradients = {}
+        for hamiltonian_parameter in self._hamiltonian.parameters:
+            summed_gradient = np.dot(
+                ansatz_gradients, self._ansatz_hamiltonian_gradients[hamiltonian_parameter]
+            )
+
+            gibbs_state_hamiltonian_gradients[hamiltonian_parameter] = summed_gradient
+
+        return gibbs_state_hamiltonian_gradients
+
+    @classmethod
+    def _discard_aux_registers(
+        cls, amplitudes_with_aux_regs: NDArray[Union[complex, float]], num_qubits_with_aux: int
```

Comment:
Can we include the information about the auxiliary registers as a property?
That way the user would always have access to this important piece of information.

Reply:
Sounds good.